### ISSUE TITLE
[iOS] ListView inside a CarouselView crashes with a NullReferenceException

### ISSUE THREAD

**Issue by velocitysystems** [xamarin/Xamarin.Forms#3039]
### Description
Place a `ListView` inside an ItemTemplate in a `CarouselView`. 

ListViewRenderer throws a NullReferenceException in one of the following `ListViewDataSource` methods under certain conditions ie. Rotate, page popped or new page pushed onto stack:
```
- GetCell(UITableView tableView, NSIndexPath indexPath)
- RowsInSection(UITableView tableview, nint section)
- NumberOfSections(UITableView tableView)
- GetHeightForHeader(UITableView tableView, nint section)
- GetViewForHeader(UITableView tableView, nint section)
- SectionIndexTitles(UITableView tableView)
```

Internally within the `ListViewDataSource` class an object is being nulled out. Seemingly, the iOS renderer for `CarouselView` is performing some kind of cleanup which is nulling out this object reference ie. During a dispose, cleanup?

A similar issue was reported [here](https://github.com/xamarin/Xamarin.Forms/issues/1992) and [here](https://github.com/alexrainman/CarouselView/issues/366). Note: The author of CarouselView @alexrainman responded with the following comment: _“This control has been merged into Xamarin.Forms master branch. Please, wait for it to be released as part of Xamarin.Forms and submit any issues there”._

Given [CarouselViewFormsPlugin](https://github.com/alexrainman/CarouselView) will soon be merged into XF, I wanted to raise this issue early so it can be resolved once the package has been merged into XF core.

### Steps to Reproduce

1. Place a `ListView` inside a `CarouselView` DataTemplate.
2. Rotate the device, pop the page or push a new page on the stack.
3. An NRE is thrown (not always, but under certain conditions).

### Expected Behavior
The app should not crash.

### Actual Behavior
The app crashes with an NRE similar to the following:
```
Object reference not set to an instance of an object (System.NullReferenceException)
  at Xamarin.Forms.Platform.iOS.ListViewRenderer+ListViewDataSource.NumberOfSections (UIKit.UITableView tableView) [0x00000] in <928eaa61bb6b442f955a09f5a4dac895>:0 
```

### Basic Information
- Version with issue: 3.0.0 and above.
- Last known good version: 2.5.x

### Workaround

- Create a custom `ListViewRenderer`.
- Create a custom `ListViewDataSource` inside the renderer which extends a `UITableViewSource`.
- Set the `Source` property to the new object.

Override affected methods and wrap call to the internal ListViewDataSource method in a try-catch:
ie. 
```
/// <summary>
/// Returns an array of titles to be displayed as an index on the table view.
/// </summary>
/// <param name="tableView">The <see cref="UITableView"/>.</param>
/// <returns>Array of titles, for example to display an alphabetized list return an array of strings from "A" to "Z".</returns>
public override string[] SectionIndexTitles(UITableView tableView)
{
    // TODO: Workaround prevents an NRE when a page containing a ListView inside a CarouselView disappears.
    // Remove once CarouselView is merged into Xamarin.Forms and this issue is fixed.
    // https://github.com/alexrainman/CarouselView/issues/366
    try
    {
        return _source.SectionIndexTitles(tableView);
    }
    catch (Exception)
    {
        return default(string[]);
    }
}
```

**Comment by alexrainman** [comment]
Placing a ListView inside the CarouselView has never been a problem, actually it has been the other way around: placing a CarouselView inside a ListView cell is source of issues.

**Comment by mattrichnz** [comment]
@alexrainman 
I'm sure that's the case, but in this case placing a `ListView` inside the `CarouselView` causes the problem. This is repeatable with the latest version of CarouselViewFormsPlugin too.

**Comment by shanwayne** [comment]
@alexrainman I'm experiencing the same issue.

**Comment by lester-fh** [comment]
@mattrichnz - I am working on building my "safelistview" renderer.  Im lost a little in your workaround.  I have my renderer coded up, but am confused where I would get _source in my safedatasource override?  Steal it from the original element in our OnElementChanged?  That renderer is pretty compact, any chance you can post your workaround in a more complete fashion?

**Comment by mattrichnz** [comment]
@lester-fh Have extract the sample workaround and put in a Gist for you.
https://gist.github.com/mattrichnz/de91e158ab1e42e33768ff8d32bcd8db

**Comment by LakMoore** [comment]
I am experiencing the same problem.  The NRE for me is always:
`GetCell(UITableView tableView, NSIndexPath indexPath)`
however, I'm not using headers or grouping in the ListView.

As an aside, I'm also having a similar problem on rotation with an OxyPlot inside CarouselView.  I'd be very cautious about adding CarouselView to XF while it is unable to host widely used Views on all platforms.  (Android works fine for me)

**Comment by LakMoore** [comment]
Fixed it!!!
I made the ListView a child of a new StackLayout and make that StackLayout the Content of the CarouselView.  I can now change the orientation of the device in iOS and Android.

**Comment by NiknakSi** [comment]
Just to add my experience of this... I've had a ListView inside a template for the carousel for ages and rotating the screen has been fine, no crashes at all. It was only when I removed the navigation bar (NavigationPage.HasNavigationBar="false") that I started to have this problem when rotating. I've used the workaround from @mattrichnz although I had to change `return default(UITableViewCell);` to `return new UITableViewCell();` since a default UITableViewCell is null and that still caused the exception for me.
